### DISCUSSION TITLE
pkce base64 url encoder

### DISCUSSION THREAD

**Discussion by wibed** [ory/hydra#2353] [Q&A] [answered]
I am having issues with the base64.RawURLEncoder encoding the sha256 digest to check the challenge against the 
code_verifier, in the pkce flow.

when I pass a random String: 
`dHB4ZWJqZXFwdW52bWJ1d2Vta3B1Z3JoZnl1Z3V0Ym8`
and encode it with the rawUrl encoder option: 
`ZEhCNFpXSnFaWEZ3ZFc1MmJXSjFkMlZ0YTNCMVozSm9abmwxWjNWMFltOA`

it checks out and encodes correctly (in the playground compared to my client)

but

when I sha256.hash above encoded random String:
`c0f27a25520e25e9a4bc7c5b4bbe6039ae98850d07a355dda9839f337b22fa7e`

even though when I `fmt.Println("%x", hash.Sum(nil))` is equivalent the above hash
when I use the same base64 encoder producing the correct output, as showed above, 

my client gets:
`YzBmMjdhMjU1MjBlMjVlOWE0YmM3YzViNGJiZTYwMzlhZTk4ODUwZDA3YTM1NWRkYTk4MzlmMzM3YjIyZmE3ZQ`

while `base64.RawURLEncoder.EncodeToString(hash.Sum(nil))` results in:
`wPJ6JVIOJemkvHxbS75gOa6YhQ0Ho1XdqYOfM3si-n4`

why does it produce the exact same output on the first example, but an entirely different one 
on the second?

I have tested it with third party options and none produces the same output as the go one, somehow

**Comment by vinckr** [comment, marked as answer]
Could you give me some pointers how to reproduce this?

**Reply by vinckr** [reply]
There was an older issue related to this: 
https://github.com/ory/hydra/issues/1572#issuecomment-565512512

**Reply by wibed** [reply]
you were right.

I should have differentiated between the raw digest and the hex encoded one. the latter my sha256 has used implicitly.
I had to include a c library to get to the raw digest.

thank you for your help. greatly appreciated

**Reply by vinckr** [reply]
No problem, glad it got cleared up quickly!
